packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.test.tsx

```diff
@@ -0,0 +1,189 @@
+  length: 3,
+  max: 10,
+  total: 40,
+};
+const graphData = {
+  areas: [
+    {
+      color: '#1CB35C',
+      dataKey: 'test-1',
+    },
+  ],
+  dimensions: [],
+  legendRowsData: [
+    {
+      data: mockMetrics,
+      format: formatPercentage,
+      handleLegendClick: vi.fn(),
+      legendColor: 'blue',
+      legendTitle: 'Test',
+    },
+  ],
+  unit: '%',
+};
+vi.spyOn(CloudPulseWidgetUtils, 'generateGraphData').mockReturnValue(graphData);
+
+class ResizeObserver {
+  disconnect() {}
+  observe() {}
+  unobserve() {}
+}
+
+const mockUpdatePreferences = vi.fn();
+vi.mock('../Utils/UserPreference', () => ({
+  useAclpPreference: () => ({
+    updateWidgetPreference: mockUpdatePreferences,
+  }),
+}));
+
+describe('Cloud pulse widgets', () => {
+  window.ResizeObserver = ResizeObserver;
+  it('should render the widget with correct title and unit', () => {
+    const { getByText } = renderWithTheme(<CloudPulseWidget {...props} />);
+    const title = getByText('CPU Utilization (%)');
+    expect(title).toBeInTheDocument();
+  });
+
+  it('should render interval select when scrape_interval is provided', () => {
+    const { getByTestId } = renderWithTheme(<CloudPulseWidget {...props} />);
+    expect(getByTestId('Data aggregation interval')).toBeInTheDocument();
+  });
+
+  it('should render aggregate function select when available_aggregate_functions exist', () => {
+    const { getByTestId } = renderWithTheme(<CloudPulseWidget {...props} />);
+    expect(getByTestId('Aggregation function')).toBeInTheDocument();
+  });
+
+  it('should render zoom icon', async () => {
+    const { getByTestId } = renderWithTheme(<CloudPulseWidget {...props} />);
+    expect(getByTestId('zoom-in')).toBeInTheDocument();
+  });
+
+  it('should render line graph component', () => {
+    const { container } = renderWithTheme(<CloudPulseWidget {...props} />);
+    expect(
+      container.querySelector('.recharts-responsive-container')
+    ).toBeInTheDocument();
+  });
+
+  it('should show error state when metrics fetch fails', () => {
+    queryMocks.useCloudPulseMetricsQuery.mockReturnValue({
+      data: undefined,
+      error: [{ reason: 'test reason' }],
+      isLoading: false,
+      status: 'error',
+    });
+
+    const { getByText } = renderWithTheme(<CloudPulseWidget {...props} />);
+    expect(getByText('test reason')).toBeInTheDocument();
+  });
+
+  it('should update preferences for zoom toggle', async () => {
+    const user = userEvent.setup();
+
+    const { getByTestId } = renderWithTheme(<CloudPulseWidget {...props} />);
+    const zoomButton = getByTestId('zoom-in');
+    await user.click(zoomButton);
+    expect(mockUpdatePreferences).toHaveBeenCalledWith('CPU Utilization', {
+      size: 6,
+    });
+  });
+
+  it('should update preferences for aggregation function select', async () => {
+    const user = userEvent.setup();
+
+    const { getByRole } = renderWithTheme(<CloudPulseWidget {...props} />);
+
+    await user.click(
```

Comment:
we can just use userEvent.click right?

Reply:
I went through some docs, setup seems to be recommended - https://github.com/testing-library/user-event/discussions/1036, https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent